docs/source/pages/developers_guide/index.rst

```diff
@@ -11,6 +11,162 @@ To follow the project's code style, see the :doc:`Macaron Style Guide </pages/de
 
 For API reference, see the :doc:`API Reference </pages/developers_guide/apidoc/index>` page.
 
+-------------------
+Writing a New Check
+-------------------
+
+As a contributor to Macaron, it is very likely to need to write a new check or modify an existing one at some point. In this
+section, we will understand how Macaron checks work and what we need to do to develop one.
+
++++++++++++++++++
+High-level Design
++++++++++++++++++
+
+Before jumping into coding, it is useful to understand how Macaron as a framework works. Macaron is an extensible
+framework designed to make writing new supply chain security analyses easy. It provides an interface
+that you can leverage to access existing models and abstractions instead of implementing everything from scratch. For
+instance, many security checks require to traverse through the code in GitHub Actions configurations. Normally,
+you would need to find the right repository and commit, clone it, find the workflows, and parse them. With Macaron,
+you don't need to do any of that and can simply write your security check by using the parsed shell scripts that are
+triggered in the CI.
+
+Another important aspect of our design is that all the check results are automatically mapped and stored in a local database.
+By performing this mapping, we make it possible to enforce flexible policies on the results of the checks. While storing
+the check results to the database happens automatically by Macaron's backend, the developer needs to add a brief specification
+to make that possible as we will see later.
+
```

Comment:
I think before elaborating on how to implement a check, we may want to clarify at a high level what a check does. Some simple high-level questions that may come to mind for a new contributor could be:
* What are the inputs of a check?
* What are the outputs of a check?
* Is it safe to assume that each check should leave no side-effects, i.e. does it only read the inputs and return the outputs, or does it modify a shared state with other checks?

Some other questions that are not as high-level but still carry a high-level understanding of the checks include:
* Do the checks run in a particular order? What is the meaning of `depends_on` and how does the check dependency relationship work?
* When are these checks executed in Macaron? What does Macaron do before and after the running of the checks? Here, it could even be helpful to emphasize that these checks run as part of `macaron analyze`.

At the same time, it could also be helpful to not just clarify what should be implemented as a check, but also what **should not** be implemented as a check, and one should leverage the policy engine instead. I feel like the `RepoCheck` example that you gave is not particularly representative of what should be implemented as a check since Macaron has already produced this information in the database and one should just write a policy instead of implementing a new check. If you think this is the case, we can either provide a different example, or simply say that "This is just an example for how a check should be implemented, and not an example of what a check should be for. Please open an issue to discuss with us if you want to implement a new check and we can confirm whether your use case is appropriate."

Reply:
Thanks for the suggestions. I will incorporate them in the documentation. A the same time, I want to keep the description short but useful.

This example check is not meant to be representative and just shows how to create a new check.  More information on what a check should not be is better to be added in a FAQ section in a different PR. Drawing a line between a check and policy and discussing what should be part of a check vs policy is also a topic for a FAQ section.

Reply:
See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).